File: CSCI-3800-Advanced-Programming/HW04/TateP_Problem3/src/Book.java
/*
    - Patrick Tate
    - CSCI 3800 HW 4 Problem 3

 */

import java.util.ArrayList; // because each book has a list of author/s

/**
 * Represent books in our library
 */
public abstract class Book implements Comparable<Book> {

    private String title;
    private String locationCode;
    private int yearPub;

    // Part 1: enable an ArrayList of Authors so a book can have multiple authors
    private ArrayList<Author> authors;
    private Publisher publisher;

    // default constructor
    public Book() {
        this.title = "";
        this.locationCode = "";
        this.yearPub = 0;
        this.authors = new ArrayList<Author>();
    }

    public Book(String title, String location, int yearPub,
                ArrayList<Author> authorList, Publisher publisher) {

        this.title = title;
        this.locationCode = location;
        this.yearPub = yearPub;

        this.publisher = publisher;

        // hw 4 enhancement
        this.authors = new ArrayList<Author>(authorList);

    }


    // Comparable
    @Override
    public int compareTo(Book o) {
        //return yearPub.compareTo(o.getYearPub());
        if (this.yearPub == o.getYearPub()) {
            return 0;
        } else if (this.yearPub > o.getYearPub()) {
            return 1;
        } else {
            return -1;
        }
    }

    // add an author the the arrayList
    public void addAuthor(Author author) {
        this.authors.add(author);
    }

    // remover an author
    // true if author in the list, false and message if not
    public boolean removerAuthor(Author author) {
        if (this.authors.contains(author)) {
            this.authors.remove(author);
            return true;
        } else {
            System.out.println("This book does not have author: " + author.getName());
            return false;
        }
    }

    // searches for a matching author string name and delete author if found
    public void deleteAuthor(String name){
        for(int i = 0; i < this.authors.size(); i++){
            if(name.equals(authors.get(i).getName())){
                authors.remove(authors.get(i));
            }
        }
    }


    // search for an author
    public boolean findAuthor(Author author) {
        if (this.authors.contains(author)) {
            System.out.println(this.title + " has author " + author.getName());
            return true;
        } else {
            System.out.println(this.title + " does not have author " + author.getName());
            return false;
        }
    }

    // return author/authors
    public ArrayList<Author> getAuthors() {
        return this.authors;
    }

    public Boolean checkTitle(String title){
        return this.title == title;
    }

    public Publisher getPublisher() {
        return publisher;
    }

    public int getYearPub() {
        return yearPub;
    }

    public void setYearPub(int yearPub) {
        this.yearPub = yearPub;
    }

    public String getLocationCode() {
        return locationCode;
    }

    public void setLocationCode(String locationCode) {
        this.locationCode = locationCode;
    }

    public String getTitle(){
        return this.title;
    }

    public void setTitle(String title) {
        this.title = title;
    }


    /**
     * The toString method converts an object to a string. It is used to display the object in a println call.
     * To be overridden in EBook and PrintedBook subclasses
     */
    public abstract String toString();

}

File: CSCI-3800-Advanced-Programming/HW04/TateP_Problem4/Book.py
# Patrick Tate
# CSCI 3800 HW 4 Problem 4

from Author import Author  # because books have an author
from Publisher import Publisher  # because books have a publisher
from abc import ABCMeta, abstractmethod  # so the Book class can be abstract

# Book class is the abstract base class that PrintedBook and ElectronicBook subclasses will be based off
# Book class has member variables: title, location, yearPub, authors(list), publisher, and pages
# Book class an abstract method PrintBook that is implemented in PrintedBook and ElectronicBook subclasses
class Book(metaclass=ABCMeta):

    # default initializer
    def __init__(self, title="", location="", yearPub=0, authors=[], publisher="", pages=0):
        super().__init__()
        self.__title = title
        self.__location = location
        self.__yearPub = yearPub
        self.__authors = []
        for author in authors:
            self.__authors.append(author)
        self.__publisher = publisher
        self.__pages = pages

    # getters
    def getTitle(self):
        return self.__title
    def getLocation(self):
        return self.__location
    def getYearPub(self):
        return self.__yearPub
    # return the author list
    def getAuthors(self):
        return self.__authors
    # return the author list as a string for easy printing
    def getAuthorsString(self):
        authorList = ""
        for author in self.__authors:
            if len(self.__authors) == 1:
                authorList = author.getName() + "."
            elif author is self.__authors[-1]:
                authorList = authorList + author.getName() + "."
            else:
                authorList = authorList + author.getName() + ", "
        return authorList
    def getPublisher(self):
        return self.__publisher
    def getPages(self):
        return self.__pages

    # setters
    def setTitle(self, title):
        self.__title = title
    def setLocation(self, location):
        self.__location = location
    def setYearPub(self, yearPub):
        self.__yearPub = yearPub
    def setAuthors(self, authors):
        self.__authors = authors
    def setPublisher(self, publisher):
        if isinstance(publisher, Publisher):
            self.__publisher = publisher
        else:
            print("*** Publisher not set: enter a Publisher object as parameter ***")
    def setPages(self, pages):
        self.__pages = pages

    # prints the author names as a string
    # if more than 1 author, author names are separated by a comma
    def printAuthors(self):
        for author in self.__authors:
            if len(self.__authors) == 1:
                print(str(author.getName()) + ".")
            elif author is self.__authors[-1]:
                print(str(author.getName()) + ".")
            else:
                print(str(author.getName()) + ", ", end="")

    # ensures the author parameter is of type Author and adds author to book if it is
    def addAuthor(self, author):
        if isinstance(author, Author):
            self.__authors.append(author)
            return True
        else:
            return False

    # ensures the author parameter is of type Author and removes author from book if it is
    def removeAuthor(self, author):
        if isinstance(author, Author):
            if author in self.__authors:
                self.__authors.remove(author)

    # returns boolean True if found
    def findAuthor(self, author):
        if isinstance(author, Author):
            if author in self.__authors:
                return True

    # compare 2 books to see which has more pages
    def compareTo(self, book):
        if isinstance(book, Book):
            if self.__pages == book.getPages():
                return 0
            elif self.__pages > book.getPages():
                return 1
            else:
                return -1

    # abstract method to be implemented in subclases
    @abstractmethod
    def printBook(self):
        pass

